README.md
[![webp-batch-convert](https://nodei.co/npm/webp-batch-convert.png)](https://npmjs.com/package/webp-batch-convert)

Webp Batch Convert
========
webp 图片批量转换。将指定目录内 png/jpg/jpeg/bmp/gif 格式的图片批量转换为 webp 格式。

[![NPM version][npm-image]][npm-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]
[![GitHub issues][issues-img]][issues-url]
[![GitHub forks][forks-img]][forks-url]
[![GitHub stars][stars-img]][stars-url]
[![minzipped size][bundlephobia-img]][bundlephobia-url]

[stars-img]: https://img.shields.io/github/stars/lzwme/webp-batch-convert.svg
[stars-url]: https://github.com/lzwme/webp-batch-convert/stargazers
[forks-img]: https://img.shields.io/github/forks/lzwme/webp-batch-convert.svg
[forks-url]: https://github.com/lzwme/webp-batch-convert/network
[issues-img]: https://img.shields.io/github/issues/lzwme/webp-batch-convert.svg
[issues-url]: https://github.com/lzwme/webp-batch-convert/issues
[npm-image]: https://img.shields.io/npm/v/webp-batch-convert.svg?style=flat-square
[npm-url]: https://npmjs.com/package/webp-batch-convert
[node-image]: https://img.shields.io/badge/node.js-%3E=_12-green.svg?style=flat-square
[node-url]: https://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/webp-batch-convert.svg?style=flat-square
[download-url]: https://npmjs.com/package/webp-batch-convert
[bundlephobia-url]: https://bundlephobia.com/result?p=webp-batch-convert@latest
[bundlephobia-img]: https://badgen.net/bundlephobia/minzip/webp-batch-convert@latest

## 快速上手

### 在项目目录中安装

```bash
npm install --save-dev webp-batch-convert
npm exec wbc -h
```

### 使用示例(nodejs 模块 API 方式)

```js
//import convert from 'webp-batch-convert';
const convert = require('webp-batch-convert');
let res;

// 示例一: 生成 img 目录下的图片文件至 webp 目录
res = await convert.cwebp('./img', './webp');
console.log('total: ', res);

// 示例二: 生成 img 目录下的图片文件至 webp 目录，附带质量等参数
// 更多参数参考：https://developers.google.com/speed/webp/docs/cwebp?csw=1#options
// 也可以执行如下命令通过 cwebp 帮助信息了解： `yarn cwebp --longhelp`
const cwebpOpts = {
    /** don't print anything */
    quiet: true,
    /** quality factor (0:small..100:big), default=75 */
    q: 75,
    /** transparency-compression quality (0..100), default=100 */
    alpha_q: 100,
    /** spatial noise shaping (0:off, 100:max), default=50 */
    sns: 50,
    /** filter strength (0=off..100), default=60 */
    f: 60,
    /** use simple filter instead of strong */
    nostrong: false,
};
// 先清空输出目录
convert.utils.delDir('./webp');
res = await convert.cwebp('./img','./webp', cwebpOpts);
console.log('total: ', res);
```

### `best-practice`

- [https://github.com/lzwme/webp-batch-convert/tree/master/best-practice](https://github.com/lzwme/webp-batch-convert/tree/master/best-practice)

## 命令行方式使用(wbc / cwebp-batch)

### 全局安装

```bash
npm install -g webp-batch-convert
wbc -h
# or 
cwebp-batch -h
```

也可以不安装，使用 `npx` 执行：

```bash
npx webp-batch-convert -h
```

### 使用示例

```js
wbc --in img-folder --out webp-folder <--debug --q 75>
```

或者局部安装，然后如下方式使用：

```js
./node_modules/.bin/wbc --in img-folder --out webp-folder <-D -q 75>
```

<p align="center">
    <img src="https://cdn.rawgit.com/lzwme/webp-batch-convert/master/test/img/snapshot.png">
</p>

## API

- `.cwebp(imgDir, webpDir, cwebpOptions)`

批量转换生成 webp。示例：
```js
// 将 img 目录下的所有图片转换为 webp 文件，输出至 webp 目录
const res = await convert.cwebp('./img','./webp', {
    debug: true,
    q: 60        // 质量
});
console.log('result: ' + res);
```

- `.utils.mkDir(dirPath)`

创建一个(深度的)目录。示例：
```js
// 创建目录
convert.utils.mkDir('./src/assets/webp');
```

- `.utils.delDir(dirPath, ext)`

清空一个（非空的）目录。示例：
```js
// 删除 webp 目录
convert.utils.delDir('./webp');
// 删除 webp 目录下的所有 webp 后缀的文件
convert.utils.delDir('./webp', 'webp');
// 删除 webp 目录下的所有 .webp、png 后缀的文件
convert.utils.delDir('./webp', /\.(webp|png)$/);
```

## 二次开发

- 依赖安装 `pnpm install`
- 修改/新增功能
- 添加测试并执行 `pnpm test`
- 全局安装与测试 `npm link . && wbc -h`

## License

`webp-batch-convert` is released under the MIT license.

该插件由[志文工作室](https://lzw.me)开发和维护。


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/lzwme/webp-batch-convert/compare/v1.0.0...v1.0.1) (2023-01-28)

## [1.0.0](https://github.com/lzwme/webp-batch-convert/compare/v0.1.4...v1.0.0) (2022-12-04)


src/cli.ts
import { program } from 'commander';
import { utils, cwebp } from './index.js';

const pkg = require('../package.json');
program
  .version(pkg.version, '-V, --version', '当前版本')
  .helpOption('-h, --help', '查看帮助信息')
  .description(pkg.description)
  .allowUnknownOption()
  // .argument('[in]', '图片源文件目录')
  // .argument('[out]', 'webp图片输出目录')
  .option('-I, --in <src>', '图片源文件目录')
  .option('-O, --out <dest>', 'webp图片输出目录')
  .option('-D, --debug', '调试模式', false)
  .option('-R, --root-dir <dirpath>', '指定程序工作目录')
  .option('-E, --extensions <ext...>', `指定要处理的图片类型`, ['jpg', 'png', 'bmp', 'jpg', 'jpeg'])
  .option('-N, --no-skip-exists', '跳过在输出目录已存在的文件', true)
  .option('-P, --parallelism <num>', '并行处理的数量。', '8')
  // ---- cwebp options start ----
  .option('-q <int>', `quality factor (0:small..100:big), default=75`)
  .option('--alpha_q <int>', `transparency-compression quality (0..100), default=100`)
  .option('-preset <string>', `preset setting.one of: default, photo, picture, drawing, icon, text`)
  .option('-z <int>', `activates lossless preset with given level in [0:fast, ..., 9:slowest]`)
  .option('-m <int>', `compression method (0=fast, 6=slowest), default=4`)
  .option('--segments <int>', `number of segments to use (1..4), default=4`)
  .option('--size <int>', `target size (in bytes)`)
  .option('--psnr <float>', `target PSNR (in dB. typically: 42)`)
  .option('-s <width...>', `input size (width x height) for YUV`)
  .option('--sns <int>', `spatial noise shaping (0:off, 100:max), default=50`)
  .option('-f <int>', `filter strength (0=off..100), default=60`)
  .option('--sharpness <int>', `filter sharpness (0:most .. 7:least sharp), default=0`)
  .option('--nostrong', `use simple filter instead of strong`)
  .option('--sharp_yuv', `use sharper (and slower) RGB->YUV conversion`)
  .option('--partition_limit <int>', `limit quality to fit the 512k limit on the first partition (0=no degradation ... 100=full)`)
  .option('--pass <int>', `analysis pass number (1..10)`)
  .option('--qrange <int...>', `<min> <max> - specifies the permissible quality range (default: 0 100)`)
  .option('--crop <int...>', `<x> <y> <w> <h> - crop picture with the given rectangle`)
  .option('--resize <int...>', `<w> <h> - resize picture (after any cropping)`)
  .option('--mt', `use multi-threading if available`)
  .option('--low_memory', `reduce memory usage (slower encoding)`)
  .option('--map <int>', `print map of extra info`)
  .option('--print_psnr', `prints averaged PSNR distortion`)
  .option('--print_ssim', `prints averaged SSIM distortion`)
  .option('--print_lsim', `prints local-similarity distortion`)
  .option('-d <file.pgm>', `dump the compressed output (PGM file)`)
  .option('--alpha_method <int>', `transparency-compression method (0..1), default=1`)
  .option('--alpha_filter <str>', `predictive filtering for alpha plane, one of: none, fast (default) or best`)
  .option('--exact', `preserve RGB values in transparent area, default=off`)
  .option(
    '--blend_alpha <hex>',
    `blend colors against background color expressed as RGB values written in hexadecimal, e.g. 0xc0e0d0 for red=0xc0 green=0xe0 and blue=0xd0`
  )
  .option('--noalpha', `discard any transparency information`)
  .option('--lossless', `encode image losslessly, default=off`)
  .option('--near_lossless <int>', `use near-lossless image preprocessing (0..100=off), default=100`)
  .option('--hint <string>', `specify image characteristics hint, one of: photo, picture or graph`)
  .option(
    '--metadata <string>',
    `comma separated list of metadata to copy from the input to the output if present. Valid values: all, none (default), exif, icc, xmp`
  )
  .option('--short', `condense printed message`)
  .option('--quiet', `don't print anything`)
  .option('--noasm', `disable all assembly optimizations`)
  .option('-v', `verbose, e.g. print encoding/decoding times`)
  .option('--progress', `report encoding progress`)
  .option('--jpeg_like', `roughly match expected JPEG size`)
  .option('--af', `auto-adjust filter strength`)
  .option('--pre', `pre-processing filter`)
  // ---- cwebp options end ----
  .action(options => {
    const src = options.in;
    const dest = options.out;
    if (options.debug) console.log(options);
    if (!src || !dest) return program.help();

    delete options.in;
    delete options.out;
    return cwebp(src, dest, options).then(r => {
      console.log(
        [
          `Successfully in \x1b[32m${r.timecost}\x1b[0m ms!`,
          ` - Total     : \x1b[96m${r.total}\x1b[0m`,
          ` - Exists    : \x1b[93m${r.exist}\x1b[0m`,
          ` - Converted : \x1b[92m${r.converted}\x1b[0m`,
        ].join('\n')
      );
    });
  });

program
  .command('rm <dirpath>')
  .description('删除一个目录及其子目录')
  .option('--ext <extension>', '按文件后缀删除文件', '.webp')
  .action((dirpath: string, options) => {
    utils.delDir(dirpath, options.ext);
  });

program.parse(process.argv);
